discovery-tasks-1 | ERROR | DefaultSchedulers | Unhandled exception (1)

```
2020-06-04 07:57:28.481+00:00 | TimeTickChannel-0 | INFO  | teku-event-log | ^[[37mSlot Event  *** Slot: 59987, Block:    ... empty, Epoch: 1874, Finalized checkpoint: 1872, Finalized root: d0278a..3bf3, Peers: 24^[[0m
2020-06-04 07:57:30.645+00:00 | discovery-tasks-1 | ERROR | DefaultSchedulers | Unhandled exception:
reactor.core.Exceptions$ErrorCallbackNotImplemented: java.util.NoSuchElementException: No value present
Caused by: java.util.NoSuchElementException: No value present
        at java.util.Optional.orElseThrow(Optional.java:375) ~[?:?]
        at org.ethereum.beacon.discovery.pipeline.handler.NodeIdToSession.getRemoteSocketAddress(NodeIdToSession.java:125) ~[discovery-0.3.6.jar:0.3.6]
        at org.ethereum.beacon.discovery.pipeline.handler.NodeIdToSession.getOrCreateSession(NodeIdToSession.java:93) ~[discovery-0.3.6.jar:0.3.6]
        at org.ethereum.beacon.discovery.pipeline.handler.NodeIdToSession.handle(NodeIdToSession.java:87) ~[discovery-0.3.6.jar:0.3.6]
        at reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.onNext(FluxPeekFuseable.java:189) ~[reactor-core-3.3.0.RELEASE.jar:3.3.0.RELEASE]
        at reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.onNext(FluxPeekFuseable.java:203) ~[reactor-core-3.3.0.RELEASE.jar:3.3.0.RELEASE]
        at reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.onNext(FluxPeekFuseable.java:203) ~[reactor-core-3.3.0.RELEASE.jar:3.3.0.RELEASE]
        at reactor.core.publisher.FluxReplay$SizeBoundReplayBuffer.replayNormal(FluxReplay.java:811) ~[reactor-core-3.3.0.RELEASE.jar:3.3.0.RELEASE]
        at reactor.core.publisher.FluxReplay$SizeBoundReplayBuffer.replay(FluxReplay.java:895) ~[reactor-core-3.3.0.RELEASE.jar:3.3.0.RELEASE]
        at reactor.core.publisher.ReplayProcessor.onNext(ReplayProcessor.java:442) ~[reactor-core-3.3.0.RELEASE.jar:3.3.0.RELEASE]
        at reactor.core.publisher.FluxCreate$IgnoreSink.next(FluxCreate.java:618) ~[reactor-core-3.3.0.RELEASE.jar:3.3.0.RELEASE]
        at reactor.core.publisher.FluxCreate$SerializedSink.next(FluxCreate.java:153) ~[reactor-core-3.3.0.RELEASE.jar:3.3.0.RELEASE]
        at org.ethereum.beacon.discovery.pipeline.PipelineImpl.push(PipelineImpl.java:43) ~[discovery-0.3.6.jar:0.3.6]
        at org.ethereum.beacon.discovery.DiscoveryManagerImpl.executeTaskImpl(DiscoveryManagerImpl.java:151) ~[discovery-0.3.6.jar:0.3.6]
        at org.ethereum.beacon.discovery.DiscoveryManagerImpl.ping(DiscoveryManagerImpl.java:162) ~[discovery-0.3.6.jar:0.3.6]
        at org.ethereum.beacon.discovery.task.LiveCheckTasks.lambda$add$2(LiveCheckTasks.java:55) ~[discovery-0.3.6.jar:0.3.6]
        at org.ethereum.beacon.discovery.scheduler.ErrorHandlingScheduler.runAndHandleError(ErrorHandlingScheduler.java:70) ~[discovery-0.3.6.jar:0.3.6]
        at org.ethereum.beacon.discovery.scheduler.ErrorHandlingScheduler.lambda$execute$1(ErrorHandlingScheduler.java:46) ~[discovery-0.3.6.jar:0.3.6]
        at org.ethereum.beacon.discovery.scheduler.Scheduler.lambda$execute$0(Scheduler.java:35) ~[discovery-0.3.6.jar:0.3.6]
        at org.ethereum.beacon.discovery.scheduler.ExecutorScheduler.lambda$execute$0(ExecutorScheduler.java:32) ~[discovery-0.3.6.jar:0.3.6]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:830) [?:?]
```

From @benjaminion https://github.com/PegaSysEng/teku/issues/2058

--- a/src/main/java/org/ethereum/beacon/discovery/pipeline/handler/NodeIdToSession.java
+++ b/src/main/java/org/ethereum/beacon/discovery/pipeline/handler/NodeIdToSession.java
@@ -84,22 +84,33 @@ public class NodeIdToSession implements EnvelopeHandler {
         "Envelope {}: Session lookup requested for nodeId {}",
         envelope.getId(),
         sessionRequest.getNodeId());
-    NodeSession nodeSession = getOrCreateSession(sessionRequest.getNodeId(), envelope);
-    envelope.put(Field.SESSION, nodeSession);
-    logger.trace("Session resolved: {} in envelope #{}", nodeSession, envelope.getId());
+    getOrCreateSession(sessionRequest.getNodeId(), envelope)
+        .ifPresentOrElse(
+            nodeSession -> {
+              envelope.put(Field.SESSION, nodeSession);
+              logger.trace("Session resolved: {} in envelope #{}", nodeSession, envelope.getId());
+            },
+            () ->
+                logger.trace(
+                    "Session could not be resolved or created for {}", sessionRequest.getNodeId()));
   }
 
-  private NodeSession getOrCreateSession(Bytes nodeId, Envelope envelope) {
-    SessionKey sessionKey = new SessionKey(nodeId, getRemoteSocketAddress(envelope));
-    NodeSession context = recentSessions.computeIfAbsent(sessionKey, this::createNodeSession);
+  private Optional<NodeSession> getOrCreateSession(Bytes nodeId, Envelope envelope) {
+    return getRemoteSocketAddress(envelope)
+        .map(
+            remoteSocketAddress -> {
+              SessionKey sessionKey = new SessionKey(nodeId, remoteSocketAddress);
+              NodeSession context =
+                  recentSessions.computeIfAbsent(sessionKey, this::createNodeSession);
 
-    sessionExpirationScheduler.put(
-        sessionKey,
-        () -> {
-          recentSessions.remove(sessionKey);
-          context.cleanup();
-        });
-    return context;
+              sessionExpirationScheduler.put(
+                  sessionKey,
+                  () -> {
+                    recentSessions.remove(sessionKey);
+                    context.cleanup();
+                  });
+              return context;
+            });
   }
 
   private NodeSession createNodeSession(final SessionKey key) {
@@ -119,10 +130,9 @@ public class NodeIdToSession implements EnvelopeHandler {
         requestExpirationScheduler);
   }
 
-  private InetSocketAddress getRemoteSocketAddress(final Envelope envelope) {
+  private Optional<InetSocketAddress> getRemoteSocketAddress(final Envelope envelope) {
     return Optional.ofNullable((InetSocketAddress) envelope.get(Field.REMOTE_SENDER))
-        .or(() -> ((NodeRecord) envelope.get(Field.NODE)).getUdpAddress())
-        .orElseThrow();
+        .or(() -> ((NodeRecord) envelope.get(Field.NODE)).getUdpAddress());
   }
 
   private static class SessionKey {

--- a/src/test/java/org/ethereum/beacon/discovery/pipeline/handler/NodeIdToSessionTest.java
+++ b/src/test/java/org/ethereum/beacon/discovery/pipeline/handler/NodeIdToSessionTest.java
@@ -10,6 +10,7 @@ import static org.mockito.Mockito.mock;
 import java.net.InetSocketAddress;
 import java.util.concurrent.Executors;
 import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.units.bigints.UInt64;
 import org.ethereum.beacon.discovery.SimpleIdentitySchemaInterpreter;
 import org.ethereum.beacon.discovery.TestUtil;
 import org.ethereum.beacon.discovery.TestUtil.NodeInfo;
@@ -17,7 +18,10 @@ import org.ethereum.beacon.discovery.pipeline.Envelope;
 import org.ethereum.beacon.discovery.pipeline.Field;
 import org.ethereum.beacon.discovery.pipeline.Pipeline;
 import org.ethereum.beacon.discovery.scheduler.ExpirationSchedulerFactory;
+import org.ethereum.beacon.discovery.schema.EnrField;
+import org.ethereum.beacon.discovery.schema.IdentitySchema;
 import org.ethereum.beacon.discovery.schema.NodeRecord;
+import org.ethereum.beacon.discovery.schema.NodeRecordFactory;
 import org.ethereum.beacon.discovery.schema.NodeSession;
 import org.ethereum.beacon.discovery.storage.AuthTagRepository;
 import org.ethereum.beacon.discovery.storage.LocalNodeRecordStore;
@@ -106,6 +110,22 @@ class NodeIdToSessionTest {
     final NodeSession session2 =
         lookupSessionForIncomingMessage(Bytes.fromHexString("0x9999"), remoteSender);
     assertThat(session1).isNotSameAs(session2);
+  }
+
+  @Test
+  void shouldNotGetASessionWhenNoAddressIsAvailable() {
+    final NodeRecord nodeRecord =
+        new NodeRecordFactory(new SimpleIdentitySchemaInterpreter())
+            .createFromValues(
+                UInt64.ONE,
+                new EnrField(EnrField.ID, IdentitySchema.V4),
+                new EnrField(EnrField.PKEY_SECP256K1, NODE_ID));
+    final Envelope envelope = new Envelope();
+    envelope.put(Field.SESSION_LOOKUP, new SessionLookup(NODE_ID));
+    envelope.put(Field.NODE, nodeRecord);
+    handler.handle(envelope);
+
+    assertThat(envelope.contains(Field.SESSION)).isFalse();
   }
 
   private NodeSession lookupSessionForIncomingMessage(

